### PR TITLE
Script engine support for Balance #983

### PR DESCRIPTION
**Pull request by AwooOOoo**
Added **balance** support into the script engine (#983). It is set up similar to _setTick_ in that it takes a callback and the trading pair to return the balance on. 

**Fields Available**
The balance exposes the following fields;
[currency=ETH, total=1000, available=1000, frozen=0, borrowed=0, loaned=0, withdrawing=0, depositing=0]

Note that the response is called twice, once for the base and once for the quote. Assuming you had ETH/EUR selected, in the example below, it would call the function once for ETH and once for EUR.

**Example Usage:**
```
function start() {
  var subscription = balance.get(
    function(event) {
      var b = event.balance();  
      notifications.info("Balance: " + b.currency + " - " + b.available);
    },
    parameters.selectedCoin
  );
  
  return RUNNING
}

function stop() {
  events.clear(subscription);  // Stop the subscription
}
```

**Output from Example:**
image

**Possible Improvements:**
1) Since it is already a balance.get() call, perhaps don't have the .balance() on the return
2) Combine returns so you can access .base() & .count() at the same time to ensure they are not out of sync
3) Make it a single call and not a subscription

Thoughts?

### CODE REVIEW

**Comment by AwooOOoo** [comment]
@badgerwithagun - I've made some mark down updates, but don't have permission to push changes and can't fork to do a PR. Whats the best way to submit them?

**Comment by badgerwithagun** [comment]
This looks great. Could we move the new method to`Events`, though, and use the same call pattern as for tickets, e.g. `setBalance`, with the assumption that it will be called again on shutdown to release the subscription?

**Comment by badgerwithagun** [comment]
One other thought, probably not one to solve here, is that the ticker and balance event objects are not version controlled DTOs, and we should probably create some for this. If those objects change, we can't have that propagating to the script API without careful consideration.

**Comment by AwooOOoo** [comment]
@badgerwithagun ,
    No worries, I'll move it to events. Assign the issue to me also. Good point on the versioning, I'll have a think about it. Also see question above about submitting documentation since wiki is a different project, I have some changes ready.

**Comment by badgerwithagun** [comment]
> @badgerwithagun Currently using Visual Studio Code. I installed eclipse, but haven't gotten into it yet and it seems to have a steep learning curve. Sorry if the PR's aren't efficient. I haven't programmed Java in many years and so am pretty damn rusty. I appreciate the hints, its a lot to take in trying to remember the language and take in the code base.

Got it. FWIW, IntelliJ is by far the easiest to use with this codebase. If you load it as a Maven project it will work out of the box with no further configuration.

**Comment by AwooOOoo** [comment]
> This is easy in IntelliJ, just right-click and run.

It is easy with IntelliJ, thanks =)

File: orko-common/src/main/java/com/gruelbox/orko/job/script/ScriptJobProcessor.java [hunk 1]
```diff
@@ -470,7 +468,35 @@ private TickerSpec convertTickerSpec(JSObject tickerSpec) {
...
   }
 
-  public Disposable onTick(
+  public class DisposableSubscription implements Disposable {
+    private ExchangeEventSubscription subscription;
+    private Disposable disposable;
+    private String description;
+
+    public DisposableSubscription(ExchangeEventSubscription subs, Disposable disp, String desc) {
+      subscription = subs;
```
Comment (badgerwithagun) [inline orko-common/src/main/java/com/gruelbox/orko/job/script/ScriptJobProcessor.java:477]: nit: the conventional pattern here is

```
this.subscription = subscription;
```

File: orko-common/src/main/java/com/gruelbox/orko/job/script/ScriptJobProcessor.java [hunk 2]
```diff
@@ -470,7 +468,35 @@ private TickerSpec convertTickerSpec(JSObject tickerSpec) {
         .build();
   }
 
-  public Disposable onTick(
+  public class DisposableSubscription implements Disposable {
+    private ExchangeEventSubscription subscription;
```
Comment (badgerwithagun) [inline orko-common/src/main/java/com/gruelbox/orko/job/script/ScriptJobProcessor.java:472]: nit: these should be final

File: orko-common/src/main/java/com/gruelbox/orko/job/script/ScriptJobProcessor.java [hunk 3]
```diff
@@ -253,6 +262,39 @@ public Disposable setTick(JSObject callback, JSObject tickerSpec) {
...
+          convertTickerSpec(tickerSpec),
+          OPEN_ORDERS,
+          callback.toString());
+    }
+
+    public Disposable setOrderBook(JSObject callback, JSObject tickerSpec) {
+      return onOrderBook(
+          event -> processEvent(() -> callback.call(null, event)),
+          convertTickerSpec(tickerSpec),
+          ORDERBOOK,
```
[the remark this answers is not in the document]
Comment (badgerwithagun) [reply orko-common/src/main/java/com/gruelbox/orko/job/script/ScriptJobProcessor.java:289]: *bump*

**Comment by badgerwithagun** [comment]
Merged. Thanks @AwooOOoo :)

**Comment by AwooOOoo** [comment]
There is a PR in the wiki, that covers the changes. I'll add the unit tests when I get some time.

**Pull request merged**